### ISSUE TITLE
Does `io_uring_submit` need to be called or not?

### ISSUE THREAD

**Issue by cmcfarlen** [axboe/liburing#579]
After upgrading to a 5.15 kernel, my io_uring application stopped working.  After staring at strace output, I realized that the `__NR_io_uring_enter` syscall was always timing out.  I added a call to `io_uring_submit` after adding new sqe entries and it started working again.  According to the `io_uring_wait_cqes` manpage:

```
 If ts is specified, the application does not need to call io_uring_submit (3) before calling io_uring_wait_cqes (3).
```

Has something recently changed where `io_uring_submit` needs to be called to flush submits to the kernel?

**Comment by axboe** [comment]
What did you upgrade from? This used to be true because the timeout was implemented as an internal request, and hence liburing would always submit for you. With newer kernels, we can pass in the timeout instead, but that does mean that you cannot rely on wait-with-timeout doing submit for you. I think this is mostly a liburing bug as it should've keep the implied submit if there are entries in the SQ ring, but it's also problematic to do so for the cases where someone safely splits submit and complete into separate threads.

In summary, yes you probably just want to ensure you submit before waiting. To not add extra syscalls, you could do ```io_uring_submit_and_wait_timeout()``` for that.

Hope that made sense...

**Comment by cmcfarlen** [comment]
I upgraded from 5.10.104.  These are the new features I noticed:

```
 IORING_FEAT_SQPOLL_NONFIXED
 IORING_FEAT_EXT_ARG
 IORING_FEAT_NATIVE_WORKERS
 IORING_FEAT_RSRC_TAGS
```

I see that `IORING_FEAT_EXT_ARG` is checked and a different codepath is taken when doing a timed wait for cqe.

**Comment by axboe** [comment]
Yep, so the addition of EXT_ARG is exactly why. As mentioned, this is really a liburing issue, it should still submit for that case, but it makes it impossible to safely use the split submit+complete approach. Wish I had a better answer for you outside of "yes the application needs to get updated because of that liburing oversight".

**Comment by cmcfarlen** [comment]
Understood.  Thank you for the quick response and updating the man pages.

**Comment by uuuuu-diwu** [comment]
It seems that using io_uring_submit_and_wait_timeout still has this problem. I originally wanted to use io_uring_submit_and_wait_timeout to try and reduce syscalls, but if io_uring_submit is not called explicitly, then io_uring_submit_and_wait_timeout will always return -ETIME, and after I call io_uring_submit, it works
